File: 257-binary-tree-paths/257-binary-tree-paths.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {string[]}
 */
const binaryTreePaths = function(root) {
    const paths = [];
    if (!root) return paths;
    
    const dfs = (node, path) => {
        path += node.val;
        if (!node.left && !node.right) {
            paths.push(path);
            return;
        }
        
        if (node.left) dfs(node.left, path + '->');
        if (node.right) dfs(node.right, path + '->');
    }
    dfs(root, '');
    
    return paths;
};

File: 257-binary-tree-paths/257-binary-tree-paths.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {String[]}
def binary_tree_paths(root)
    dfs root, res = [], []
    res
end

def dfs(node, res, path)
    return res << [*path, node.val].join('->') if (!node.left && !node.right)
    
    dfs node.left, res, [*path, node.val] if node.left
    dfs node.right, res, [*path, node.val] if node.right
end